flink-mesos/src/main/java/org/apache/flink/mesos/runtime/clusterframework/MesosResourceManagerDriver.java

```diff
@@ -0,0 +1,713 @@
+	}
+
+	/**
+	 * Extracts the Mesos task goal state from the worker information.
+	 *
+	 * @param worker the persistent worker information.
+	 * @return goal state information for the {@link TaskMonitor}.
+	 */
+	private static TaskMonitor.TaskGoalState extractGoalState(MesosWorkerStore.Worker worker) {
+		switch(worker.state()) {
+			case New: return new TaskMonitor.New(worker.taskID());
+			case Launched: return new TaskMonitor.Launched(worker.taskID(), worker.slaveID().get());
+			case Released: return new TaskMonitor.Released(worker.taskID(), worker.slaveID().get());
+			default: throw new IllegalArgumentException("unsupported worker state");
+		}
+	}
+
+	private static String extractTerminatedDiagnostics(ResourceID id, Protos.TaskStatus status) {
+		return String.format("Worker %s terminated with status: %s, reason: %s, message: %s.",
+				id, status.getState(), status.getReason(), status.getMessage());
+	}
+
+	private static TaskSchedulerBuilder createOptimizer() {
+		return new TaskSchedulerBuilderImpl();
+	}
+
+	private void assertStateCleared() {
+		assert(workersInNew.isEmpty());
+		assert(requestResourceFutures.isEmpty());
+	}
+	// ------------------------------------------------------------------------
+	//  Internal Classes
+	// ------------------------------------------------------------------------
+
+	private class MesosResourceManagerSchedulerCallback implements Scheduler {
+
+		@Override
+		public void registered(SchedulerDriver driver, final Protos.FrameworkID frameworkId, final Protos.MasterInfo masterInfo) {
+			getMainThreadExecutor().execute(() -> MesosResourceManagerDriver.this.registered(new Registered(frameworkId, masterInfo)));
+		}
+
+		@Override
+		public void reregistered(SchedulerDriver driver, final Protos.MasterInfo masterInfo) {
+			getMainThreadExecutor().execute(() -> MesosResourceManagerDriver.this.reregistered(new ReRegistered(masterInfo)));
+		}
+
+		@Override
+		public void resourceOffers(SchedulerDriver driver, final List<Protos.Offer> offers) {
+			getMainThreadExecutor().execute(() -> MesosResourceManagerDriver.this.resourceOffers(new ResourceOffers(offers)));
+		}
+
+		@Override
+		public void offerRescinded(SchedulerDriver driver, final Protos.OfferID offerId) {
+			getMainThreadExecutor().execute(() -> MesosResourceManagerDriver.this.offerRescinded(new OfferRescinded(offerId)));
+		}
+
+		@Override
+		public void statusUpdate(SchedulerDriver driver, final Protos.TaskStatus status) {
+			getMainThreadExecutor().execute(() -> MesosResourceManagerDriver.this.statusUpdate(new StatusUpdate(status)));
+		}
+
+		@Override
+		public void frameworkMessage(SchedulerDriver driver, final Protos.ExecutorID executorId, final Protos.SlaveID slaveId, final byte[] data) {
+			// noop
+		}
+
+		@Override
+		public void disconnected(SchedulerDriver driver) {
+			getMainThreadExecutor().executeWithoutFencing(() -> MesosResourceManagerDriver.this.disconnected(new Disconnected()));
```

Comment:
Looks like changing `executeWithoutFencing` to `execute` does not break any test cases.

I can try to provide an implementation that handles disconnected messages with fencing. Could you tell me how to run it on `flink-jepsen`?

Reply:
I think you cannot do it alone because we have the Jepsen source code for running it on Travis in a private repo.

I started the jepsen run with this branch: https://github.com/tillrohrmann/flink/tree/disableUnfencedExecutionMesos

We can see the CI run here: https://travis-ci.com/github/dataArtisans/flink-jepsen-ci/builds/189052564

Once the run passes I will open a PR for the changes and then merge it. You can already rebase this PR on my branch if you want.